### PR TITLE
Fix #17: adds basic audio player functionality

### LINKED ISSUE #17: Full UI: AudioFragment [Blocked: #162]
## Explanation
There needs to be a system for streaming and playing audio packs selected for a specific chapter by the user, leveraging the default audio language set for the current user's profile.

## Low-fi Checklist
- [x] Basic audio player UI 
- [x] Alert dialog for cellular network warning (https://xd.adobe.com/view/bee1fced-33bd-4f69-6a89-2c30fe1c4f65-c8d0/screen/d74f0571-471f-4ea6-a3a5-32e135253c56/Lesson-Introduction-4)
- [ ] Based on audio, content-card should get highlighted
- [x] Audio language selection
- [x] Play audio via MediaPlayer or ExoPlayer (https://github.com/google/ExoPlayer)

## High-fi Checklist
- [x] Seekbar and its animation
Screenshot 2019-11-08 at 3 24 11 PM

- [x] AudioBar along with Volume icon in toolbar
Screenshot 2019-11-08 at 3 24 37 PM

- [x] CellularAlertDialog
Screenshot 2019-11-08 at 3 24 58 PM

- [x] LanguageChangeDialog
Screenshot 2019-11-08 at 3 24 45 PM

- [x] Final touch including animations, positioning and Accessibility checks

## Test follow items too
- [x] Check how to audio bar changes on `en` and `hi-en` because this might change the seekbar length.
- [x] Shadow below the seekbar thumb (green circle)
- [x] The seekbar progress - white line has a sort of padding at start/end which needs to be removed.
- [x] The seekbar white progress line should be rounded at the end
- [ ]After the volume icon click, the “Language string like `HI` or `EN`” takes time to be visible which is not good for the UI.

### PR DESCRIPTION
**Pull request by jamesxu0**
## Explanation

This PR contains an implementation of a basic audio player with temporary UI. The audio player will be added onto Rajat's work in the exploration player after reviewed.

## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is **assigned** to an appropriate reviewer.

### CODE REVIEW

**Comment by BenHenning** [comment]
Hi @jamesxu0. When you request a review from someone, please also assign the issue to them so that they know to take action on the PR. Similarly, when the review is completed the PR will be assigned back to you to take action (either resolve comments or submit the PR when it's ready).

File: domain/src/main/java/org/oppia/domain/AudioPlayerController.kt
Comment (BenHenning) [inline domain/src/main/java/org/oppia/domain/AudioPlayerController.kt]: This ought to be in an audio subpackage. The top-level package should just be a container of more specific subpackages.

File: build.gradle
Comment (BenHenning) [inline build.gradle]: Is there a specific reason you're updating the Gradle version? Prefer not making that change with other code changes in case there are downstream implications.

File: domain/src/main/java/org/oppia/domain/AudioPlayerController.kt
Comment (BenHenning) [inline domain/src/main/java/org/oppia/domain/AudioPlayerController.kt]: So we need to model this in a way where we don't have a listener. Ensuring listeners are properly added/removed within the correct lifecycle boundaries is difficult, and is the reason mechanism like ``LiveData`` were introduced.

Could we represent this controller using a ``DataProvider`` like ``UserAppHistoryController``? Consider something like:

```kotlin
audioPlayerController.playFromUrl(audioUrl)
audioPlayerController.getPlayStatusLiveData().observe(this) { playStatus ->
   // ... playStatus includes information about the readiness, position, duration, and completion state, or anything the UI needs.
}
```

Let me know if you want to chat about this more.

File: app/src/main/java/org/oppia/app/home/HomeActivityController.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/home/HomeActivityController.kt]: I don't think we want to ship this. Maybe add a TODO that this needs to be reverted.

File: app/src/main/java/org/oppia/app/audioplayer/AudioPlayerFragmentController.kt
Comment (BenHenning) [inline app/src/main/java/org/oppia/app/audioplayer/AudioPlayerFragmentController.kt]: Nit: this should be a 'Presenter' not a 'Controller'.

Also, how should this interop with the one Rajat is introducing in #105? Is it still needed with that PR?

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @jamesxu0--I had a few high-level questions, and I suspect this is also blocked on #105. I'd like to do a deep-dive review on this once the high-level questions are addressed.

**Comment by rt4914** [comment]
@jamesxu0 I have checked this PR and there are changes but rather than describing it here I think they are more descriptive so maybe have a look at #105 so that you can have an idea.

**Review by rt4914 — COMMENTED** [review]
There are changes to this PR but in general I suggest you to have a look at #105 .

Expected Major Changes:
1. Use data-binding, MVVM concepts
2. Your UI, especially the language selection does not match with the Mocks. We do have to create everything but also we have keep UI in mind and work accordingly.
3. I guess we can have a discussion on whether we want to use Exoplayer or MediaPlayer for audio. Can you write an email describing which one is advantages in why.

**Comment by jamesxu0** [comment]
@BenHenning PTAL. I changed the implementation to use live data to keep track of the play status. I did not have this problem yesterday, but now i get this error when loading audio "java.io.FileNotFoundException: No content provider". I only get this when running the app through an emulator, but it's fine on a real device. It may have to do with the network connection type.

File: app/src/main/java/org/oppia/app/audioplayer/AudioPlayerFragmentController.kt
Comment (jamesxu0) [reply app/src/main/java/org/oppia/app/audioplayer/AudioPlayerFragmentController.kt]: I think I would just add upon #105 to include listeners that would interact with the AudioPlayerController

File: domain/src/main/java/org/oppia/domain/AudioPlayerController.kt
Comment (jamesxu0) [reply domain/src/main/java/org/oppia/domain/AudioPlayerController.kt]: Done. Added audio package

File: build.gradle
Comment (jamesxu0) [reply build.gradle]: Done. Accidental change

File: domain/src/main/java/org/oppia/domain/AudioPlayerController.kt
Comment (jamesxu0) [reply domain/src/main/java/org/oppia/domain/AudioPlayerController.kt]: I updated the AudioPlayerController to expose LiveData to the fragment to observe on.

File: app/src/main/java/org/oppia/app/home/HomeActivityController.kt
Comment (jamesxu0) [reply app/src/main/java/org/oppia/app/home/HomeActivityController.kt]: Done. reverted, just used for testing purposes

**Comment by jamesxu0** [comment]
This PR shouldn't be merged. Once approved, I'll use part of this implementation and integrate with #105

**Comment by BenHenning** [comment]
Per my recent email, let's update this once #130 and #123 are finished. You can send this back out for review before #116 is done, but we should wait on submitting it until after since it's finalizing the bar (so everything ought to be working after this PR goes in).

**Comment by jamesxu0** [comment]
Pull request out of date, audio player is now being worked on in PR #149

**Pull request closed without merging**
